# Assignment 4:
    test if the distribution of ages is the same for 
    
    1) male and female riders
    
    2) day and night riders
    
    use 3 tests: KS, Pearson's, Spearman's. 
    
    For the KS do the test with the entire dataset and with a subset 200 times smaller
    
    Choose a significant threshold for the whole eercise. 
    
    For each test phrase the Null Hypothesis in words.
    
    Describe the return of the scipy function you use in each case.
    
    State the result in terms of rejection of the Null.

In [43]:
# my usual imports and setups
import pylab as pl
import pandas as pd
import numpy as np
#imports downloader
from getCitiBikeCSV import getCitiBikeCSV

%pylab inline

import os

Populating the interactive namespace from numpy and matplotlib


# I am reading in data from January 2015. It would be a good idea to use data from warmer months, since there are more riders in the warm weather and the more data the smaller the STATISTICAL NOISE. You should use data from multiple months, joining multiple datasets, thus addressing systematic errors as well

In [44]:
datestring = '201505'
getCitiBikeCSV(datestring)

file in place, you can continue


In [45]:
df = pd.read_csv(os.getenv("PUIDATA") + "/" + datestring + '-citibike-tripdata.csv')
df.head()

tripduration          starttime           stoptime  start station id  \
0           415  5/1/2015 00:00:11  5/1/2015 00:07:06               477   
1          1523  5/1/2015 00:00:28  5/1/2015 00:25:52               293   
2           642  5/1/2015 00:00:52  5/1/2015 00:11:35               380   
3           367  5/1/2015 00:00:48  5/1/2015 00:06:55               537   
4          2734  5/1/2015 00:00:51  5/1/2015 00:46:26               426   

        start station name  start station latitude  start station longitude  \
0          W 41 St & 8 Ave               40.756405               -73.990026   
1    Lafayette St & E 8 St               40.730287               -73.990765   
2         W 4 St & 7 Ave S               40.734011               -74.002939   
3  Lexington Ave & E 24 St               40.740259               -73.984092   
4    West St & Chambers St               40.717548               -74.013221   

   end station id          end station name  end station latitude  \
0             442           W 27 St & 7 Ave             40.746647   
1             324   DeKalb Ave & Hudson Ave             40.689888   
2             507           E 25 St & 2 Ave             40.739126   
3             280           E 10 St & 5 Ave             40.733320   
4             327  Vesey Pl & River Terrace             40.715338   

   end station longitude  bikeid    usertype  birth year  gender  
0             -73.993915   17012  Subscriber      1981.0       1  
1             -73.981013   17390    Customer         NaN       0  
2             -73.979738   15003  Subscriber      1990.0       1  
3             -73.995101   14788  Subscriber      1978.0       1  
4             -74.016584   21068  Subscriber      1956.0       2

In [46]:
df['date'] = pd.to_datetime(df['starttime'])
df.head()

tripduration          starttime           stoptime  start station id  \
0           415  5/1/2015 00:00:11  5/1/2015 00:07:06               477   
1          1523  5/1/2015 00:00:28  5/1/2015 00:25:52               293   
2           642  5/1/2015 00:00:52  5/1/2015 00:11:35               380   
3           367  5/1/2015 00:00:48  5/1/2015 00:06:55               537   
4          2734  5/1/2015 00:00:51  5/1/2015 00:46:26               426   

        start station name  start station latitude  start station longitude  \
0          W 41 St & 8 Ave               40.756405               -73.990026   
1    Lafayette St & E 8 St               40.730287               -73.990765   
2         W 4 St & 7 Ave S               40.734011               -74.002939   
3  Lexington Ave & E 24 St               40.740259               -73.984092   
4    West St & Chambers St               40.717548               -74.013221   

   end station id          end station name  end station latitude  \
0             442           W 27 St & 7 Ave             40.746647   
1             324   DeKalb Ave & Hudson Ave             40.689888   
2             507           E 25 St & 2 Ave             40.739126   
3             280           E 10 St & 5 Ave             40.733320   
4             327  Vesey Pl & River Terrace             40.715338   

   end station longitude  bikeid    usertype  birth year  gender  \
0             -73.993915   17012  Subscriber      1981.0       1   
1             -73.981013   17390    Customer         NaN       0   
2             -73.979738   15003  Subscriber      1990.0       1   
3             -73.995101   14788  Subscriber      1978.0       1   
4             -74.016584   21068  Subscriber      1956.0       2   

                 date  
0 2015-05-01 00:00:11  
1 2015-05-01 00:00:28  
2 2015-05-01 00:00:52  
3 2015-05-01 00:00:48  
4 2015-05-01 00:00:51

In [47]:
df.columns

Index(['tripduration', 'starttime', 'stoptime', 'start station id',
       'start station name', 'start station latitude',
       'start station longitude', 'end station id', 'end station name',
       'end station latitude', 'end station longitude', 'bikeid', 'usertype',
       'birth year', 'gender', 'date'],
      dtype='object')

# BY GENDER
splitting data by gender and extracting the age happens in the cell below

In [48]:
#df is the dataframe where the content of the csv file is stored
df['ageM'] = 2015 - df['birth year'][(df['usertype'] == 'Subscriber') & (df['gender'] == 1)]
df['ageF'] = 2015 - df['birth year'][(df['usertype'] == 'Subscriber') & (df['gender'] == 2)]

In [49]:
df.head(5)

tripduration          starttime           stoptime  start station id  \
0           415  5/1/2015 00:00:11  5/1/2015 00:07:06               477   
1          1523  5/1/2015 00:00:28  5/1/2015 00:25:52               293   
2           642  5/1/2015 00:00:52  5/1/2015 00:11:35               380   
3           367  5/1/2015 00:00:48  5/1/2015 00:06:55               537   
4          2734  5/1/2015 00:00:51  5/1/2015 00:46:26               426   

        start station name  start station latitude  start station longitude  \
0          W 41 St & 8 Ave               40.756405               -73.990026   
1    Lafayette St & E 8 St               40.730287               -73.990765   
2         W 4 St & 7 Ave S               40.734011               -74.002939   
3  Lexington Ave & E 24 St               40.740259               -73.984092   
4    West St & Chambers St               40.717548               -74.013221   

   end station id          end station name  end station latitude  \
0             442           W 27 St & 7 Ave             40.746647   
1             324   DeKalb Ave & Hudson Ave             40.689888   
2             507           E 25 St & 2 Ave             40.739126   
3             280           E 10 St & 5 Ave             40.733320   
4             327  Vesey Pl & River Terrace             40.715338   

   end station longitude  bikeid    usertype  birth year  gender  \
0             -73.993915   17012  Subscriber      1981.0       1   
1             -73.981013   17390    Customer         NaN       0   
2             -73.979738   15003  Subscriber      1990.0       1   
3             -73.995101   14788  Subscriber      1978.0       1   
4             -74.016584   21068  Subscriber      1956.0       2   

                 date  ageM  ageF  
0 2015-05-01 00:00:11  34.0   NaN  
1 2015-05-01 00:00:28   NaN   NaN  
2 2015-05-01 00:00:52  25.0   NaN  
3 2015-05-01 00:00:48  37.0   NaN  
4 2015-05-01 00:00:51   NaN  59.0

In [50]:
#your df should look something like this
# look at these data carefully... you may see someinteresting values!
df.describe()

/opt/rh/anaconda/root/envs/PUI2016_Python3/lib/python3.5/site-packages/numpy/lib/function_base.py:3403: RuntimeWarning: Invalid value encountered in median
  RuntimeWarning)


tripduration  start station id  start station latitude  \
count  9.619860e+05     961986.000000           961986.000000   
mean   1.000234e+03        442.963114               40.734483   
std    6.828864e+03        322.369486                0.019656   
min    6.000000e+01         72.000000               40.646768   
25%    4.130000e+02        306.000000               40.720828   
50%    6.750000e+02        410.000000               40.736197   
75%    1.131000e+03        490.000000               40.750073   
max    2.937702e+06       2023.000000               40.771522   

       start station longitude  end station id  end station latitude  \
count            961986.000000   961986.000000         961986.000000   
mean                -73.991070      441.553910             40.734237   
std                   0.012412      321.888012              0.019626   
min                 -74.017134       72.000000             40.646768   
25%                 -74.000264      306.000000             40.720664   
50%                 -73.990931      408.000000             40.735439   
75%                 -73.981948      489.000000             40.749718   
max                 -73.950048     2023.000000             40.771522   

       end station longitude         bikeid     birth year         gender  \
count          961986.000000  961986.000000  794434.000000  961986.000000   
mean              -73.991170   18107.945644    1976.250738       1.018968   
std                 0.012427    2066.054842      11.458816       0.606306   
min               -74.017134   14529.000000    1885.000000       0.000000   
25%               -74.001497   16344.000000            NaN       1.000000   
50%               -73.990931   18107.000000            NaN       1.000000   
75%               -73.981948   19841.000000            NaN       1.000000   
max               -73.950048   21907.000000    1999.000000       2.000000   

                ageM           ageF  
count  607999.000000  186113.000000  
mean       39.139096      37.468662  
std        11.510284      11.130245  
min        16.000000      17.000000  
25%              NaN            NaN  
50%              NaN            NaN  
75%              NaN            NaN  
max       116.000000     115.000000

In [51]:
# dropping NaN values
df['ageM'].dropna(inplace= True)
df['ageF'].dropna(inplace= True)

lets split age in 10 year bins. the dataset is very large, so i could be split in smaller bins, but I will chose 10 years in the interest of time. 
the bin size choice should be a balance between properly sample the age space, have enough counts in each bin that the statistical noise is not significant (remember that is > sqrt(N)!) and the computational requirement to computatinal facilities ratio. 

the next several steps are needed if you want to code up the KS test from scratch. that is for extra credit, so if you do not want to do it you may not need to plot split the distribution in bins and create the cumulative HOWEVER it is a great idea to do it anyways to explore your data viaually! remember Ascombe's quartet!!

In [53]:
bins = np.arange(10, 99, 5)
df.ageM.groupby(pd.cut(df.ageM, bins)).agg([count_nonzero]).cumsum()

count_nonzero
ageM                   
(10, 15]            0.0
(15, 20]         6081.0
(20, 25]        54122.0
(25, 30]       163926.0
(30, 35]       277091.0
(35, 40]       360768.0
(40, 45]       434378.0
(45, 50]       496302.0
(50, 55]       547189.0
(55, 60]       578929.0
(60, 65]       597593.0
(65, 70]       604427.0
(70, 75]       607292.0
(75, 80]       607672.0
(80, 85]       607719.0
(85, 90]       607723.0
(90, 95]       607768.0

In [54]:
df.ageM.groupby(pd.cut(df.ageM, bins)).agg([count_nonzero])

count_nonzero
ageM                   
(10, 15]            0.0
(15, 20]         6081.0
(20, 25]        48041.0
(25, 30]       109804.0
(30, 35]       113165.0
(35, 40]        83677.0
(40, 45]        73610.0
(45, 50]        61924.0
(50, 55]        50887.0
(55, 60]        31740.0
(60, 65]        18664.0
(65, 70]         6834.0
(70, 75]         2865.0
(75, 80]          380.0
(80, 85]           47.0
(85, 90]            4.0
(90, 95]           45.0

In [ ]:
# plotting histogramswith pandas is a bitdifferent than with pylab
# this is a VERY USEFUL syntaxfor you to knonw!
bins = np.arange(10, 99, 5)
axM = df.ageM.groupby(pd.cut(df.ageM, bins)).agg([count_nonzero]).plot(kind='bar', 
                                                                legend=False)
axM.set_title("male riders")
axF = df.ageF.groupby(pd.cut(df.ageF, bins)).agg([count_nonzero]).plot(kind='bar',
                                                                legend=False)
axF.set_title("female riders")


# Figure 1: histogrammed distribution of riders' ages by gender

this is how the cumulative distributions look like.  Notice that i am normalizing them! if i want to reat an observed distribution like a probablility distribution i have to normalize it

In [ ]:
#print df.ageS, df.ageS.cumsum()

csM=df.ageM.groupby(pd.cut(df.ageM, bins)).agg([count_nonzero]).cumsum()

csF=df.ageF.groupby(pd.cut(df.ageF, bins)).agg([count_nonzero]).cumsum()

print (np.abs(csM / csM.max()-csF / csF.max()))

pl.plot(bins[:-1] + 5, csM / csM.max(), label = "M")
pl.plot(bins[:-1] + 5, csF / csF.max(), label = "F")
pl.plot(bins[:-1] + 5, np.sqrt(csF / csF.max() - csM / csM.max())**2, 'k-',
        label = "difference")
pl.xlabel("Age")
pl.ylabel("Normalized Cumulative Number")
pl.legend()

# Figure 2: the cumulative distribution of CitiBike riders' ages by gender

They look similar! But the difference gets to 10%. If I wanted to code the KS test by hand I woud have everything I need: the normalized cumulative distributions can be subtracted from each other and the max distance can calculated. 

Notice that there may be NaN values you are gonna have to deal with! 
You can do that for example with a Boolean statementsuch as  df.ageF[~np.isnan(df.ageF)] or you can use numpy functions that deal with Nan values: nansum, nanmean, nanstd...

lets run the scipy KS test

In [55]:
import scipy.stats

# KS tests to compare 2 samples

http://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.ks_2samp.html


## Null Hypothesis: There is no correlation between the age of female and male riders
### Significance Level: 5%

In [56]:
ks_age = scipy.stats.ks_2samp(df.ageM, df.ageF)
print (ks_age)

Ks_2sampResult(statistic=0.079410589622185057, pvalue=0.0)


### Calculate $c(\alpha) \sqrt{\frac{(n1 + n2)}{n1n2}}$

($c(\alpha)$  is 1.36 at the significance level of 5%)

In [57]:
np.sqrt((len(df.ageM) + len(df.ageF))/(len(df.ageM)*len(df.ageF)))*1.36

0.0036027987255167846

### The KS test returns two values: the KS statistic and the p-value. Since the KS statistic is larger than $c(\alpha) \sqrt{\frac{(n1 + n2)}{n1n2}}$ (0.794 > 0.0036) or the p-value is below 5% (~0.0%), we can reject the Null hypothesis    


# Now redo the test with a subsample of the data: take only 1 ride every of 200 

In [62]:
#Create 2 sub sample for male and female
ageM_sub = df.ageM.sample(frac=1/200, replace=True)
ageF_sub = df.ageF.sample(frac=1/200, replace=True)

In [61]:
ks_age_sub = scipy.stats.ks_2samp(ageM_sub, ageF_sub)
print (ks_age_sub)

Ks_2sampResult(statistic=0.069636143931256722, pvalue=0.0018667065849998061)


### The KS test returns two values: the KS statistic and the p-value. Since the KS statistic is larger than $c(\alpha) \sqrt{\frac{(n1 + n2)}{n1n2}}$ (0.0696 > 0.0036) or the p-value is below 5% (~0.2%), we can reject the Null hypothesis    

# Pearson's  test for correlation

## notice that the Pearson's is a pairwise test: the samples need to be 
## a. sorted and 
## b. the same size!

http://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.pearsonr.html#scipy.stats.pearsonr



In [79]:
ageM_pearson = np.random.choice(df.ageM, size=len(df.ageF), replace=False, p=None)
ageF_pearson = np.array(df.ageF)

In [95]:
ageM_pearson.sort()
ageF_pearson.sort()

In [111]:
# your code here!
# wrangle the data as needed
# please perform the Pearson's test 
# and tell me what you find in terms of NULL hypothesis

Pearson = scipy.stats.pearsonr(ageM_pearson, ageF_pearson)
print(Pearson)

(0.99563985048417591, 0.0)


 ### The Pearson's test returns the Pearson correlation coefficient and the p-value: 0.9956, 0.0 respectively. The results indicate a strong positive correlation between the age of male and female riders with the coefficient close to 1. Since the p-value is lower than our significance level of 5% (~0.0%), we can reject our Null hypothesis. 

# Spearman's  test for correlation

http://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.spearmanr.html#scipy.stats.spearmanr

In [112]:
# your code here!
# wrangle the data as needed
ageM_spearman = np.random.choice(df.ageM, size=len(df.ageF), replace=False, p=None)
ageM_spearman.sort()
ageF_spearman = ageF_pearson
Spearman = scipy.stats.spearmanr(ageM_spearman, ageF_spearman)
print(Spearman)

SpearmanrResult(correlation=0.99905502316247907, pvalue=0.0)


### The Spearman's test return the rank-order correlation coefficient and the p-value, which are 0.999 and 0.0 respectively. The result indicate a strong positive correlation between the age of male and female riders with the coefficient close to 1. Since the p-value is lower than our significance level of 5% (~0.0%), we can reject our Null hypothesis. The Spearman's test result is consistent with the Pearson's test result with a low p-value and small variance in the coefficient value.

## please comment on the numbers that you get in the light of the scipy manual: what is rerutned? what does it mean??

## EXTRA CREDIT:  CHOOSE ONE OF THE TESTS, AND TEST THE CORRELATION OF AGES  OF RIDERS DURING THE DAY VS DURING THE NIGHT!



In [137]:
df['age_night'] = 2015 - df['birth year'][(df['usertype'] == 'Subscriber') & (df.date.dt.hour > 19) & (df.date.dt.hour <=23)
                                          & (df.date.dt.hour < 7)]
df['age_day'] = 2015 - df['birth year'][(df['usertype'] == 'Subscriber') & (df.date.dt.hour >= 7) & (df.date.dt.hour <= 19)]

In [138]:
df['age_night'].dropna(inplace= True)
df['age_day'].dropna(inplace= True)

## KS test
### Null Hypothesis: There is no correlation between the age of day and night riders
Significance Level: 5%

In [139]:
ks_age_time = scipy.stats.ks_2samp(df.age_night, df.age_day)
print (ks_age)

Ks_2sampResult(statistic=0.079410589622185057, pvalue=0.0)


### The KS test returns two values: the KS statistic and the p-value, 0.0794 and 0.0 respectively. Since the KS statistic is larger than $c(\alpha) \sqrt{\frac{(n1 + n2)}{n1n2}}$ (0.0794 > 0.0036) or the p-value is below 5% (~0.2%), we can reject the Null hypothesis that the age distributions of the day and night riders are the same    